### PR TITLE
vtls: fix potential version string overflow

### PR DESCRIPTION
**Pull request by danielgustafsson**
In `Curl_multissl_version()` it was possible to overflow the passed in buffer if the generated version string exceeded the size of the buffer. Fix by inverting the logic, and also make sure to not exceed the local buffer during the string generation.

Reported-by: nevv on HackerOne/curl

### CODE REVIEW

File: lib/vtls/vtls.c [hunk 1]
```diff
@@ -1243,7 +1243,7 @@ static size_t Curl_multissl_version(char *buffer, size_t size)
 
     selected = current;
 
-    for(i = 0; available_backends[i]; i++) {
+    for(i = 0; available_backends[i] && p < backends + sizeof(backends); i++) {
```
Comment (jay) [inline lib/vtls/vtls.c]: You'll have to account for each time p is incremented. How about
~~~c
  if(current != selected) {
    char *p = backends;
    char *end = backends + sizeof(backends);
    int i;

    selected = current;

    for(i = 0; available_backends[i] && p < end - 3; i++) {
      if(i)
        *(p++) = ' ';
      if(selected != available_backends[i])
        *(p++) = '(';
      p += available_backends[i]->version(p, end - p - 3);
      if(selected != available_backends[i])
        *(p++) = ')';
    }
    *p = '\0';
    total = p - backends;
  }
~~~
Also I suggest increase it to like 1k size, 200 seems low for multiple versions.

File: lib/vtls/vtls.c [hunk 2]
```diff
@@ -1256,14 +1256,14 @@ static size_t Curl_multissl_version(char *buffer, size_t size)
...
-  if(size < total)
+  if(size > total)
     memcpy(buffer, backends, total + 1);
   else {
     memcpy(buffer, backends, size - 1);
     buffer[size - 1] = '\0';
   }
 
-  return total;
+  return CURLMIN(size, total);
```
Comment (jay) [inline lib/vtls/vtls.c]: size - 1

File: lib/vtls/vtls.c (hunk 1 above)
Comment (jay) [reply lib/vtls/vtls.c]: also what is supposed to happen here if(current==selected)?

Comment (bagder) [reply lib/vtls/vtls.c]: If we need to increase the `backends` buffer, or perhaps independently of that, we should also increase the `version` and the `ssl_version` buffers that are used to get this output (in lib/version.c) - both of them are probably in the smaller end if someone builds a curl with all possible TLS backends enabled (not sure exactly how many and which that might be).

Comment (danielgustafsson) [reply lib/vtls/vtls.c]: > also what is supposed to happen here if(current==selected)?

It's keeping the buffer cached in case the current TLS backend hasn't changed between calls, as the selectdt backend is printed differently from the rest.

Comment (danielgustafsson) [reply lib/vtls/vtls.c]: > You'll have to account for each time p is incremented.

Good catch, fixing.

> Also I suggest increase it to like 1k size, 200 seems low for multiple versions.

I'm not sure 200 characters is too low, and 1k seems like excessive stack usage, but there are definitely buffers in the version printing which needs to get bumped. That's will be a separate PR though.

File: lib/vtls/vtls.c (hunk 2 above)
Comment (danielgustafsson) [reply lib/vtls/vtls.c]: Ah, yes.

**Comment by danielgustafsson** [comment]
Pushed a rebased fixup version with the comments addressed.

File: lib/vtls/vtls.c (hunk 1 above)
Comment (bagder) [reply lib/vtls/vtls.c]: I would imagine 200 should be enough. If you manage to build with **6** backends, that makes it 33 bytes per backend, and I don't think any backend uses that much. **8** backends give 25 bytes each, probably also plenty enough.

I doubt we can build with more than 8, if even that...

Comment (jay) [reply lib/vtls/vtls.c]: > Good catch, fixing.

I've added `p += available_backends[i]->version(p, end - p - 2);` since it wasn't in the fix. Technically if bytes written returned is always going to be less than the count passed (eg version(foo, 5) and therefore the max it will return is 4 bec it accounts for the null) you only need to subtract 1 but I think subtracting 2 makes the intention clearer.

**Comment by danielgustafsson** [comment]
Squashed and pushed this to master, thanks for review!

File: lib/vtls/vtls.c [hunk 3]
```diff
@@ -1239,31 +1239,32 @@ static size_t Curl_multissl_version(char *buffer, size_t size)
...
     selected = current;
 
-    for(i = 0; available_backends[i]; i++) {
+    for(i = 0; available_backends[i] && p < (end - 4); i++) {
       if(i)
         *(p++) = ' ';
       if(selected != available_backends[i])
         *(p++) = '(';
-      p += available_backends[i]->version(p, backends + sizeof(backends) - p);
+      p += available_backends[i]->version(p, end - p - 2);
```
Comment (dscho) [inline lib/vtls/vtls.c]: Shouldn't this be an `end - p > 2 ? end - p - 2 : 0`? The second parameter is declared as `size_t`, i.e. unsigned...

Comment (jay) [reply lib/vtls/vtls.c]: end - p is always > 2 at that point, refer to the loop condition `p < (end - 4)`, then after that `p` can be incremented twice before `end - p - 2`

Comment (danielgustafsson) [reply lib/vtls/vtls.c]: Right, unless I'm missing something @jay is right, the current coding should be safe (but it might warrant a comment explaining it).

Comment (jay) [reply lib/vtls/vtls.c]: I find adding a comment to be unnecessary I think it's pretty straightforward.

Comment (jzakrzewski) [reply lib/vtls/vtls.c]: Well, one could just right the loop condition as `...&& (end - p) > 4` and it'd be easier to read

Comment (dscho) [reply lib/vtls/vtls.c]: > I find adding a comment to be unnecessary I think it's pretty straightforward.

So you're saying that I am dumb. Because I did not see that.

I don't take this personally, coming from you, as I had pleasant and productive interactions with you in the past, but I wonder whether you realized how unfortunate a reaction your comment would elicit (and I think it is pretty obvious that it would do exactly that). Might want to be more careful.

In short: while it might be obvious to you, *now*, I think that my comment is testament enough to give evidence to the contrary. And who knows, maybe you yourself will belong into the same "wait, what?" camp in 6 months from now.

Comment (bagder) [reply lib/vtls/vtls.c]: I'm in @dscho's corner here. When I first saw the question here about this line of code it was not immediately obvious to me that it was right. A comment could've remedied that both for @dscho and for me. For those who don't need the comment, it won't hurt (much) anyway.

Comment (jay) [reply lib/vtls/vtls.c]: > > I find adding a comment to be unnecessary I think it's pretty straightforward.
> 
> So you're saying that I am dumb. Because I did not see that.
> 
> I don't take this personally, coming from you, as I had pleasant and productive interactions with you in the past, but I wonder whether you realized how unfortunate a reaction your comment would elicit (and I think it is pretty obvious that it would do exactly that). Might want to be more careful.
>
> In short: while it might be obvious to you, _now_, I think that my comment is testament enough to give evidence to the contrary. And who knows, maybe you yourself will belong into the same "wait, what?" camp in 6 months from now.

Whoa I may not be very tactful but don't put words in my mouth. I figured you misread the code, like I do occasionally. I would've shared my opinion even if that wasn't the case. If you had a different opinion that's all you needed to say in your reply.

Anyway, clearly my opinion of it being straightforward is not in favor. Someone propose a comment so we can put this to bed. `/* end - p is > 2 at this point */` ?
